{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Understanding figure 1c\n",
    "This is a notebook that breaks down figure 1c in the original article https://www.nature.com/articles/s41586-020-2907-3. There won't be codes in here, just cells with codeblocks.\n",
    "\n",
    "Disclaimer: This is written by a person from outside neuroscience who had nothing to do with Scala et al.'s research. Some contents may  be wrong.\n",
    "\n",
    "## Table of contents:\n",
    "* [Summary of the original article by Scala et al.](#article_summary)\n",
    "* [The article's description about figure 1c](#article_figure1c)\n",
    "* [The plotting code for figure 1c](#plotting_figure1c)\n",
    "* [The t-SNE process and rnaseqTools.map_to_tsne](#about_tsne)\n",
    "* [About the datasets](#datasets)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Summary of the original article by Scala et al. <a class=\"anchor\" id=\"article_summary\"></a>\n",
    "\n",
    "According to the article by Scala et al., there are two ways to classify neuron types: transcriptomic and  morpho-electric. The transcriptomic method uses the RNA sequence of neurons to determine their families. The morpho-electric (morphology and electrphysiology) method first observes actual neuron shapes by coloring them with a substance called byoctin. Next it runs electrical currents through the neurons and records the reactions of the neurons. Finally, it combines the neuron shapes to their reactions under electricity to classify the cells into families.\n",
    "\n",
    "Scala's article combined both transcriptomic and  morpho-electric descriptions to mouse neurons from the primary motor cortex (MOp) to see what it could find about existing research on  transcriptomic neuron classification. The experiment showed that the morpho-electric classification agreed with the transcriptomic classification for broad cell families (Pvalb, Vip, Lamp5, and other code-like names found in the article). However, morpho-electric classification did not match transcriptomic classification for smaller neuron subclasses. The article concludes that the smaller neuron families may not be distinct and discrete as the transcriptomic classification suggests, but is more like a spectrum within a large group. \n",
    "\n",
    "When assigning transcriptomic cell types, Scala et al. referenced two transcriptomic cell data sets, one from Tasic et al. 2018 and one from Yao et al. 2020. Scala et al. first used Tasic et al.'s data set to assign initial categories to the cell samples they had. Then they did the final assignment with Yao et al.'s 7 datasets for RNA sequences.\n",
    "\n",
    "### The referenced datasets\n",
    "\n",
    "Yao et al. classify neurons in the primary motor cortex using RNA and DNA sequencing and chromatin. For figure 1c Scala et al. only use one of the 7 datasets for the RNA sequence for assigning transciptomic cell types.<br>\n",
    "https://www.biorxiv.org/content/10.1101/2020.02.29.970558v2.full.pdf\n",
    "\n",
    "Tasic et al. classified 23,822 neurons from mice by single-cell RNA sequencing. These neurons were from two areas of mouse brains: the primary visual cortex and the anterior lateral motor cortex.<br>\n",
    "https://www.nature.com/articles/s41586-020-2907-3#Abs1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. The article's description about figure 1c <a class=\"anchor\" id=\"article_figure1c\"></a>\n",
    "\n",
    "According to the article, 1c is: \n",
    ">t-SNE representation of CGE-derived interneurons from the single-cell 10x v2 reference data set (n = 15,511; perplexity, 30).\n",
    " \n",
    "From what can be gathered, the main data is from the \"single-cell 10x v2 reference data set\". The dataset contained a set of interneurons that came from a place called CGE, and in the figure, the data is represented by something called a t-SNE.\n",
    "\n",
    "To someone with no background, the questions are:\n",
    "1. What is an interneuron?\n",
    "2. Where and what is the CGE?\n",
    "3. What is a t-SNE representation?\n",
    "4. Is there any meaning to the dataset's name, and where did it come from?\n",
    "\n",
    "__1. What is an interneuron?__\n",
    "\n",
    "It is a common type of neuron that can be found in the brain and spinal cord. They pass signals through the nervous system, which allows us to sense and move.\n",
    "\n",
    "__2. Where and what is the CGE?__\n",
    "\n",
    "It is a tiny region of a brain section called the telencephalon, which is associated with the motor functions (movements) of the body. Along with the CGE, there are sections called the LGE and MGE which are also part of the telencephalon. CGE is the second largest source of cortical interneurons (number one source is MGE).\n",
    "\n",
    "__3. What is a t-SNE?__\n",
    "\n",
    "It is a non-linear dimensionality reduction technique developed in 2008 that is effective for high-dimensional data (to compare, PCA is a linear dimensionality reduction technique). It works by using a Gaussian distribution to calculate the conditional probability of one point choosing another point as a close neighbor. It does this for the original data and for a lower dimension representation, and then uses KL divergence to decide what low dimension representation best fits the original data. It has a parameter \"perplexity\", which controls the \"how close is a close neighbor\" aspect. Like PCA, you end up with a set of unidentifiable features that are complex combinations of the original. \n",
    "\n",
    "For details, look at the following:\n",
    "- https://www.datacamp.com/community/tutorials/introduction-t-sne\n",
    "- https://distill.pub/2016/misread-tsne/\n",
    "- https://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf\n",
    "\n",
    "__4. Is there any meaning to the data set's name, and where did it come from?__\n",
    "\n",
    "Single cell sequencing is the technology of identifying a DNA or RNA sequence of one cell, and 10x sequencing is an advanced method that lets you do it. This dataset in particular, the \"single-cell 10x v2 reference data set\", is from Yao et al.'s article in 2020."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. The plotting code for figure 1c <a class=\"anchor\" id=\"plotting_figure1c\"></a>\n",
    "\n",
    "Combining the information from the article and the actual code from ttype-coverage-mod.ipynb, you get detailed information on what exactly figure 1c is.\n",
    "\n",
    "It is a t-SNE mapping of one group of neurons, the CGE interneurons, in one of Yao at al.'s datasets that represents the results of single cell 10x v2 RNA sequencing. This dataset contains classifications of 15,511 cells and can be found in `10X_cells_v2_AIBS.pickle`, a pickle file you can obtain by running `allen-data-preprocess-mod.ipynb`.\n",
    "\n",
    "In figure 1c, clusters created by the t-SNE algorithm are colored by the actual transcriptomic cell types Yao at al. classified. Broad cell families like Vip, Lamp5, and Sncg are also plotted. Additionally, approximate locations of where Scala et al.'s cells from the same area would be are overlayed by black markers. The marker's shape indicates what level of the brain the cell was found in.\n",
    "\n",
    "The code is shown with additional comments below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The code with comments\n",
    "\n",
    "----\n",
    "\n",
    "```python\n",
    "Z = pickle.load(open('../data/processed/rnaseq/10x-tsne-viplamp.pickle', 'rb')) # load t-SNE representations of Yao et al.'s Vip/Lamp5 cells\n",
    "subset = np.isin(ttypes['family'], ['Lamp5', 'Vip']) # a mask that gets cells that were assigned \"Lamp5\" or \"Vip\" in Scala et al.'s samples\n",
    "\n",
    "# assign t-SNE representations to Scala et al.'s cells (m1) using Yao et al.'s cells (m1data, Z is t-SNE features) as reference\n",
    "# refer to the next section in this notebook if you want to know more about it\n",
    "pos = rnaseqTools.map_to_tsne(m1data['viplamp']['counts'], m1data['viplamp']['genes'],\n",
    "                              m1.exonCounts[subset,:], m1.genes, Z, knn=10,\n",
    "                              newIntronCounts=m1.intronCounts[subset,:], \n",
    "                              newExonLengths=m1.exonLengths, newIntronLengths=m1.intronLengths,\n",
    "                              normalizeNew=True, normalizeReference=False)\n",
    "\n",
    "# create a dictionary of t-SNE representations of Scala et al.'s cells, keys are areas of brain\n",
    "# this bit of code is not used when plotting figure 1c\n",
    "pos_all = {}\n",
    "pos_all['viplamp'] = pos\n",
    "                              \n",
    "# this block plots a scatterplot of Yao et al.'s cells, in colors specified in Yao et al.'s article\n",
    "# the colors can be found in 10X_cells_v2_AIBS.pickle\n",
    "plt.sca(ax3)\n",
    "plt.gca().set_aspect('equal', adjustable='datalim')\n",
    "plt.scatter(Z[:,0], Z[:,1], s=1, alpha=1, rasterized=True, edgecolors='none',\n",
    "            c = clusterColors[m1data['viplamp']['clusters']])\n",
    "\n",
    "# overlay Scala et al.'s 244 cells with Yao et al.'s clusters of 15,511 CGE interneurons\n",
    "# the estimated location of cells were derived in rnaseqTools.map_to_tsne()\n",
    "# the marker shapes and sizes represent the layers in which the cells were found\n",
    "sc = []\n",
    "for L in ['1', '2/3', '5', '6']:\n",
    "    s = plt.scatter(pos[layers[subset]==L,0], pos[layers[subset]==L,1], \n",
    "                    s=sizes[L], c='k', zorder=2, marker=shapes[L])\n",
    "    sc.append(s)\n",
    "    \n",
    "leg = plt.legend(sc, ['L1','L2/3','L5','L6'], fontsize=5, loc='upper right')\n",
    "leg.get_frame().set_linewidth(.5)\n",
    "\n",
    "# from here the code plots the t-type names for the clusters\n",
    "# offsets: the offset [x,y] to avoid overlap with dots on the scatterplot\n",
    "# angles: set angle to make tilted captions\n",
    "# renames: if you  need to set a new name for the t-types, do it here\n",
    "offsets = {'Lamp5 Slc35d3': [25,-20], 'Lamp5 Lhx6': [10,6], 'Lamp5 Pdlim5_2': [20,10],\n",
    "           'Lamp5 Pdlim5_1': [20,5], 'Lamp5 Pax6': [10,8], 'Lamp5 Egln3_1': [20,8],\n",
    "           'Lamp5 Egln3_2': [15,7], 'Lamp5 Egln3_3': [25,16], 'Vip Igfbp6_1': [15,-10],\n",
    "           'Vip Igfbp6_2': [15,-5], 'Vip C1ql1': [5,-20], 'Vip Gpc3': [0,-12],\n",
    "           'Vip Chat_1': [0,-17], 'Vip Mybpc1_2': [-10,15], 'Vip Mybpc1_1': [-8,5],\n",
    "           'Sncg Slc17a8': [0,9], 'Sncg Calb1_1': [-8,7], 'Sncg Calb1_2': [-12,3],\n",
    "           'Sncg Npy2r': [-15,0], 'Vip Sncg': [-12,5], 'Sncg Col14a1': [7,0],\n",
    "           'Vip Htr1f': [15,-9], 'Vip Serpinf1_3': [2,-10], 'Vip Serpinf1_1': [-4,3],\n",
    "           'Vip Mybpc1_3': [-10,0], 'Vip Serpinf1_2': [10,0], 'Vip Chat_2': [0,-10]}\n",
    "\n",
    "angles = {'Vip Mybpc1_1': 60, 'Sncg Col14a1': -75}\n",
    "renames = {'Vip Mybpc1_3': 'M3', 'Vip Serpinf1_1': 'S1', 'Vip Serpinf1_2': 'S2', 'Vip Serpinf1_3': 'S3'}\n",
    "\n",
    "for c in np.unique(m1data['viplamp']['clusters']): # for every cluster assigned to Yao et al.'s CGE interneurons\n",
    "    ind = m1data['viplamp']['clusters']==c\n",
    "    col = clusterColors[c]\n",
    "    x,y = np.median(Z[ind,0]), np.median(Z[ind,1]) #set the labels at the median value of the t-SNE values for that cluster\n",
    "    if clusterNames[c] in offsets: # use the offset\n",
    "        x += offsets[clusterNames[c]][0]\n",
    "        y += offsets[clusterNames[c]][1]\n",
    "    if clusterNames[c] not in renames:\n",
    "        label = '\\n'.join(clusterNames[c].split()[1:]) #for brevity, keep the cell family name out and replace space with newline\n",
    "    else:\n",
    "        label = renames[clusterNames[c]] #rename if you need to\n",
    "    if clusterNames[c] not in angles:\n",
    "        alpha = 0\n",
    "    else:\n",
    "        alpha = angles[clusterNames[c]] #set angles to cell types if specified\n",
    "        \n",
    "    if ~np.isnan(x):\n",
    "        plt.text(x, y, label, color=col, fontsize=5, ha='center', va='center', zorder=1, rotation=alpha)\n",
    "\n",
    "#the next 6 rows are for plotting the cell family names: Vip, Sncg, Lamp5\n",
    "col = clusterColors[clusterNames=='Vip Mybpc1_3'][0]\n",
    "plt.text(-65, -50, 'Vip', color='w', fontsize=6, bbox=dict(boxstyle='round', ec=col, fc=col))\n",
    "col = clusterColors[clusterNames=='Sncg Calb1_2'][0]\n",
    "plt.text(-65, 55, 'Sncg', color='w', fontsize=6, bbox=dict(boxstyle='round', ec=col, fc=col))\n",
    "col = clusterColors[clusterNames=='Lamp5 Slc35d3'][0]\n",
    "plt.text(50, -60, 'Lamp5', color='w', fontsize=6, bbox=dict(boxstyle='round', ec=col, fc=col))\n",
    "\n",
    "#these remove ticks from the plot and erases borderline of plot\n",
    "plt.xticks([])\n",
    "plt.yticks([])\n",
    "sns.despine(ax=ax3, left=True, bottom=True)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. The t-SNE process and rnaseqTools.map_to_tsne <a class=\"anchor\" id=\"about_tsne\"></a>\n",
    "\n",
    "Actually, the real t-SNE process does not happen when plotting figure 1c. The `rnaseqTools.map_to_tsne` function used when plotting figure 1c is for assigning estimated t-SNE mappings from representations that were already calculated elsewhere.\n",
    "\n",
    "This part is for understanding the t-SNE process used in the article and what goes on when `rnaseqTools.map_to_tsne` is called when plotting figure 1c.\n",
    "\n",
    "### The t-SNE process\n",
    "The exact method is described in the Method section in Scala et al.'s article (https://www.nature.com/articles/s41586-020-2907-3#Sec7). This section also gives a reference to another article that invented the protocol: https://www.nature.com/articles/s41467-019-13056-x.\n",
    "1. Normalize the count data so that the UMI counts can be compared across all cells, each with different sequence depths. The normalization method is `log2(UMI_counts_of_cell / seqdepth_of_cell * median(seqdepth_of_allcells) + 1)`\n",
    "2. Do PCA on the normalized counts, so that the features are reduced from `number_of_cells * number_of_genes` to `number_of_cells * 50`\n",
    "3. Perform t-SNE on the feature-reduced counts using a tool from https://github.com/KlugerLab/FIt-SNE. The tool maps to 2 dimensions unless otherwise specified"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  rnaseqTools.map_to_tsne with comments\n",
    "\n",
    "As stated before, this is not the t-SNE itself, but a function for assigning estimated t-SNE values from reference t-SNE representations. It works by calculating the correlation between the UMI counts or exon and intron counts of the reference cells and new cells, and then assigns the median t-SNE representation of the reference cells with highest correlation to the new cells.\n",
    "\n",
    "---\n",
    "\n",
    "\n",
    "```python\n",
    "def map_to_tsne(referenceCounts, referenceGenes, newCounts, newGenes, referenceAtlas, \n",
    "                bootstrap = False, knn = 10, nrep = 100, seed = None, batchsize = 1000,\n",
    "\t\t\t\tverbose = 1,\n",
    "                referenceIntronCounts = None, newIntronCounts = None,\n",
    "                normalizeNew = False, normalizeReference = False,\n",
    "                newExonLengths = None, newIntronLengths = None):\n",
    "    \"\"\"\n",
    "    This function assigns t-SNE representations to cells by \n",
    "    1. Calculating the normalized exonic + intronic expression levels for new cells\n",
    "    2. Getting their correlations to the reference counts\n",
    "    3. Assigning the median t-SNE representation of the reference cells highest correlation\n",
    "    4. Bootstrapping if specified\n",
    "    \n",
    "    The arguments are:\n",
    "    - referenceCounts: The exon counts for reference cells. If the counts are not separated into exons and introns, it is used for UMI counts. Figure 1c uses the UMI counts for 15,511 Vip/Lamp5 cells from Yao et al.\n",
    "    - referenceGenes: The gene name list for reference cells\n",
    "    - newCounts: The exon counts for the new cells that need to be assigned t-SNE representations. If the counts are not separated into exons and introns, it is used for UMI counts. Figure 1c uses the 244 Vip/Lamp5 cells from Scala et al.\n",
    "    - newGenes: The names of genes that were detected in the new cells\n",
    "    - referenceAtlas: The t-SNE mappings of the reference cells. Figure 1c uses the mapping of 15,511 Yao's Vip/Lamp5 cells\n",
    "    - bootstrap: If True does bootstrap\n",
    "    - knn: A parameter that is used for getting top n correlations\n",
    "    - nrep: The number of times bootstrapping is repeated\n",
    "    - seed: Random seed for bootstrapping\n",
    "    - batchsize: Batchsize for processing data\n",
    "    - verbose: If larger than zero, prints progress\n",
    "    - referenceIntronCounts: The intron counts for reference cells. If the counts are not separated into exons and introns, remains None.\n",
    "    - newIntronCounts: The intron counts for the new cells. If the counts are not separated into exons and introns, remains None. Figure 1c uses the 244 Vip/Lamp5 cells from Scala et al.\n",
    "    - normalizeNew: If True, counts for the new cells are normalized with exon/intron lengths\n",
    "    - normalizeReference: If True, counts for the reference cells are normalized with exon/intron lengths\n",
    "    - newExonLengths: The list of exon lengths. Figure 1c uses the data for 42,466 genes from Scala et al.\n",
    "    - newIntronLengths: The intron lengths. Figure 1c uses the data for 42,466 genes from Scala et al.\n",
    "    \n",
    "    return:\n",
    "    - bootstrap is True: The estimated t-SNE assignments of the new cells, a tuple of (non-bootstrapped version, bootstrapped version)\n",
    "    - bootstrap is False: The non-bootstrapped version of estimated t-SNE assignments of the new cells\n",
    "    \"\"\"\n",
    "    \n",
    "    # get a list of genes which can be found in the gene list for both reference and new cells\n",
    "    # and sort them in alphabetical order\n",
    "    gg = sorted(list(set(referenceGenes) & set(newGenes)))\n",
    "    if verbose > 0:\n",
    "        print('Using a common set of ' + str(len(gg)) + ' genes.')\n",
    "    \n",
    "    newGenes = [np.where(newGenes==g)[0][0] for g in gg] # the indices in new gene list that correspond to the common genes\n",
    "    refGenes = [np.where(referenceGenes==g)[0][0] for g in gg] # the indices in reference gene list that correspond to the common genes\n",
    "    \n",
    "    ### Step 1-1. Get normalized exonic + intronic expression levels for new cells ###\n",
    "    \n",
    "    X = newCounts[:,newGenes] # get the exon counts of the common genes from new cells\n",
    "    if sparse.issparse(X): \n",
    "        X = np.array(X.todense()) # if X is a sparse array, convert it to a normal numpy array\n",
    "    if normalizeNew:\n",
    "        X = X / (newExonLengths[newGenes]/1000) # normalize exon counts by corresponding exon lengths (normalization by kilobase)\n",
    "    # do the same for the intron counts\n",
    "    if newIntronCounts is not None:\n",
    "        Xi = newIntronCounts[:, newGenes]\n",
    "        if sparse.issparse(Xi):\n",
    "            Xi = np.array(Xi.todense())\n",
    "        if normalizeNew:\n",
    "            Xi = Xi / ((newIntronLengths[newGenes]+.001)/1000) #add a very small number to avoid division by zero\n",
    "        X = X + Xi # lastly, add add the intron counts and exon counts together to create normalized exonic + intronic expression levels\n",
    "    X = np.log2(X + 1) #log transform\n",
    "    \n",
    "    ### Step 1-2. Get normalized exonic + intronic expression levels for reference cells ###\n",
    "    \n",
    "    # do the same thing for reference cells.\n",
    "    T = referenceCounts[:,refGenes]\n",
    "    if sparse.issparse(T):\n",
    "        T = np.array(T.todense())\n",
    "    if normalizeReference:\n",
    "        T = T / (newExonLengths[newGenes]/1000)\n",
    "    if referenceIntronCounts is not None:\n",
    "        Ti = referenceIntronCounts[:, refGenes]\n",
    "        if sparse.issparse(Ti):\n",
    "            Ti = np.array(Ti.todense())\n",
    "        if normalizeReference:\n",
    "            Ti = Ti / ((newIntronLengths[newGenes]+.001)/1000)\n",
    "        T = T + Ti\n",
    "    T = np.log2(T + 1)\n",
    "    \n",
    "    ### Step 2 & 3. Get highest correlation and assign median t-SNE representations ###\n",
    "    \n",
    "    n = X.shape[0] #number of new cells\n",
    "    assignmentPositions = np.zeros((n, referenceAtlas.shape[1])) # array for t-SNE assignment. size num_new_cells * t-SNE_output_dim\n",
    "    batchCount = int(np.ceil(n/batchsize)) # get number of batches\n",
    "    if (batchCount > 1) and (verbose > 0):\n",
    "        print('Processing in batches', end='', flush=True) \n",
    "    for b in range(batchCount):\n",
    "        if (batchCount > 1) and (verbose > 0):\n",
    "            print('.', end='', flush=True) \n",
    "        batch = np.arange(b*batchsize, np.minimum((b+1)*batchsize, n)) # this just creates indices for each batch\n",
    "        C = corr2(X[batch,:], T) # get correlation matrix of new cells and reference\n",
    "        ind = np.argpartition(C, -knn)[:, -knn:] # get indices of 10 largest-ish correlations\n",
    "        for i in range(batch.size):\n",
    "            # assign the median t-SNE representation of the reference cells with10 largest-ish correlations\n",
    "            assignmentPositions[batch[i],:] = np.median(referenceAtlas[ind[i,:],:], axis=0)\n",
    "    if (batchCount > 1) and (verbose > 0):\n",
    "        print(' done', flush=True) \n",
    "    \n",
    "    ### Step 4. Do bootstrapping if specified ###\n",
    "    \n",
    "    # Note: currently bootstrapping does not support batchsize\n",
    "    if bootstrap:\n",
    "        if seed is not None:\n",
    "            np.random.seed(seed)\n",
    "        assignmentPositions_boot = np.zeros((n, referenceAtlas.shape[1], nrep))\n",
    "        if verbose>0:\n",
    "            print('Bootstrapping', end='', flush=True)\n",
    "        for rep in range(nrep):\n",
    "            if verbose>0:\n",
    "                print('.', end='')\n",
    "            bootgenes = np.random.choice(T.shape[1], T.shape[1], replace=True)\n",
    "            C_boot = corr2(X[:,bootgenes],T[:,bootgenes])\n",
    "            ind = np.argpartition(C_boot, -knn)[:, -knn:] \n",
    "            for i in range(X.shape[0]):\n",
    "                assignmentPositions_boot[i,:,rep] = np.median(referenceAtlas[ind[i,:],:], axis=0)\n",
    "        if verbose>0:\n",
    "            print(' done')      \n",
    "        return (assignmentPositions, assignmentPositions_boot)\n",
    "    else:\n",
    "        return assignmentPositions\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. About the datasets <a class=\"anchor\" id=\"datasets\"></a>\n",
    "\n",
    "There are three variables that have datasets stored in them: `ttypes`, `m1`, and `m1data`. `ttypes`and `m1` are different cuts of Scala et al.'s dataset, and `m1data` is part of Yao et al.'s dataset.\n",
    "\n",
    "### About ttypes\n",
    "\n",
    "It contains information about the category assignments of 1,329 cells that were used as samples in Scala et al.'s article. Details can be found in the Methods section, under \"t-Type assignment\".\n",
    "\n",
    "- __qc:__ Quality control result\n",
    "- __tasicCorrs:__ Pearson correlation with Tasic et al.'s cell types\n",
    "- __tasic_ass:__ Classification to Tasic et al.'s cell types. Basically select the one with higest correlation but cells with `qc = low quality` are assigned `nan`\n",
    "- __tasic_ass_boot :__ The bootstrap result (proportion) used in assignment. The bootstrap selected 100 sets of random genes from a cell with repetition, saw what type the sets were assigned to, and selected the type with the highest proportion of assignment\n",
    "- __m1ass:__ Classification to Yao et al.'s cell types. Separated by the seven reference datasets\n",
    "- __m1corrs:__ Pearson correlation with Yao et al.'s cell types. Separated by the seven reference datasets\n",
    "- __m1boot:__ Classification to Yao et al.'s cell types. Separated by the seven reference datasets\n",
    "- __m1consensus_boot:__ The bootstrap result (proportion) used in the final assignment. This is a majority vote across all 7 reference datasets\n",
    "- __m1consensus_ass:__ The assignment result according to the bootstrap across all datasets. In other words, `argmax(m1consensus_boot)` for every cell.\n",
    "- __type:__ The names of the cell types each cell was assigned to\n",
    "- __family:__ The broad cell family each cell was assigned to\n",
    "- __confidence:__ The highest value in the bootstrap proportions, or in other words, the `m1consensus_boot` value corresponding to the assigned cell type\n",
    "\n",
    "### About m1\n",
    "\n",
    "It contains the results of the study Scala et al. did on their 1,329 cells. It has the full list of detected genes, so the gene number is 42,466.\n",
    "\n",
    "- __exonCounts:__ The exon count for each cell and gene. Exons are sections of a gene. Size `number_of_cells * number_of_genes`.\n",
    "- __intronCounts:__ The intron count for each cell and gene. Introns are also sections of a gene. Size`number_of_cells * number_of_genes`.\n",
    "- __cells:__ Name of each cell\n",
    "- __genes:__ Names of all 42,466 genes that were detected\n",
    "- __layer:__ Layer of primary motor cortex in which cell was found\n",
    "- __cre:__ Cre line of mouse the cell came from. Kind of like a bloodline of genetically engineered lab mice.\n",
    "- __yields:__ (Probably) the concentration of RNA sampled from the cell into a fluid used for tesing\n",
    "- __traced:__ Whether the cell's morphology could be recorded. 646 cells were recorded in total\n",
    "- __depth:__ How deep the soma (neuron's body) could be found in the brain. Probably larger the deeper\n",
    "- __thickness:__ The cortical thickness of the area corresponding to that cell. Cortical thickness is the thickness of the brain's outer layers, where the cells were sampled from.\n",
    "- __ephys:__ 29 electrophysiological features for each cell. Size `number_of_cells * number_of_features`\n",
    "- __ephysNames:__ The names of the 29 features\n",
    "- __exclude:__ Contains reasons for exclusion for cells that were excluded for analysis. Cells that were not excluded have the value `''`\n",
    "- __mice_ages:__ The ages of the 266 mice used in the experiment. In days\n",
    "- __mice_cres:__ The Cre line of each mouse\n",
    "- __morphometrics:__ The 61 recorded morphometric features for each cell. The features will be a list of nans for cells that couldn't be measured\n",
    "- __morphometricsNames:__ Names of the 61 morphometric features\n",
    "- __zProfiles:__ (Probably) a histogram with 20 bins per cell that indicates the mass of the neuron located at each depth. Size `number_of_cells * number_of_bins`\n",
    "- __exonLengths:__ The length of the exon for each gene\n",
    "- __intronLengths:__ The length of the intron for each gene\n",
    "\n",
    "### About m1data\n",
    "It is a section of the dataset10X_cells_v2_AIBS.pickle, where only the CGE interneurons were extracted. The original pickle file is a dictionary whose first layer keys are the subgroups: CGE interneurons (viplamp), MGE interneurons (pvsst), excitatory neurons (exc), and all groups put together (neurons). The keys each have another dictionary associated with it, which has info about the subgroups:\n",
    "\n",
    "- __counts:__ The UMI count for each gene and cell, a sparse matrix of size `number_of_cells * number_of_genes`. UMI counts are the number of observed transcirpts\n",
    "- __genes:__ The gene list. This only includes 1,000 most variable genes for the neuron group and 500 most variable genes for the subgroups.\n",
    "- __clusters:__ The cluster assignments for each cell. The cluster types are from Yao et al.'s article and stored in a file called `cluster.annotation.csv`. 10X_cells_v2_AIBS only uses the 88 types where `class_label != \"Other\"`.\n",
    "- __clusterColors:__ A color assigned to each cluster. Used for plotting. This one contains the same information regardless of group.\n",
    "- __clusterNames:__ A name assigned  to each cluster.\n",
    "- __seqDepths:__ How many genes were read per cell. Not equal to the sum of counts, perhaps because it included genes that were not in the gene list.\n",
    "\n",
    "And just a sidenote, the remaining subgroups in the 10X_cells_v2_AIBS dataset, the MGE interneurons and excitatory cells, are mapped in figure 1d and 1e as similar plots respectively. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
